src/Graph/Trees/Heap.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+namespace Rubix\ML\Graph\Trees;
+
+use InvalidArgumentException;
+
+/**
+ * Heap
+ *
+ * A heap is a specialized tree-based data structure that satisfies the heap property:
+ * in a max heap, for any given node C, if P is a parent node of C, then the key of P
+ * is greater than or equal to the key of C. In a min heap, the key of P is less than or equal to the key of C.
+ *
+ * References:
+ * [1] Black (ed.), Paul E. (2004-12-14)
+ * <https://xlinux.nist.gov/dads/HTML/heap.html>
+ *
+ * @category    Machine Learning
+ * @package     RubixML
+ * @author      Rich Davis
+ */
+class Heap
+{
+    /**
+     * An array of words from a corpus arranged in a manner that maintains the heap invariant.
+     *
+     * @var mixed[]
+     */
+    protected $heap;
+
+    /**
+     * @param array[] $x
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(array $x)
+    {
+        if (!is_array($x)) {
+            throw new InvalidArgumentException('Heap requires a valid array.');
+        }
+
+        $this->heap = $this->heapify($x);
+    }
+
+    /**
+     * Returns the heap.
+     *
+     * @return array[]
+     */
+    public function heap() : array
+    {
+        return $this->heap;
+    }
+
+    /**
+     * Pops and returns the smallest item in the heap.
+     *
+     * @return mixed[]|null
+     */
+    public function heappop() : ?array
```

Comment:
Since this a priority queue, "pop" will return the smallest item and while maintaining the heap invariant. "Pop" and "extract" are somewhat synonymous but we would have to explicitly state "extract-min" in this situation. I referenced python's priority queue naming convention which reads "heappop".

```
heapq.heappop(heap)
Pop and return the smallest item from the heap, maintaining the heap invariant. If the heap is empty, IndexError is raised. To access the smallest item without popping it, use heap[0].
```

https://docs.python.org/3/library/heapq.html

Reply:
ah gotcha, handn't grokked initially that this was a priority queue, and not a generic (min|max)heap structure 👍